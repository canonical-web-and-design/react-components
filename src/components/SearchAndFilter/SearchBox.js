import classNames from "classnames";
import React from "react";
import PropTypes from "prop-types";

const SearchBox = React.forwardRef(
  (
    {
      autocomplete = "off",
      className,
      disabled,
      onChange,
      onSubmit,
      placeholder = "Search",
      value,
      ...props
    },
    ref
  ) => {
    const submit = (e) => {
      e.preventDefault();
      onSubmit && onSubmit(e);
    };
    return (
      <form
        className={classNames("p-search-and-filter__box", className)}
        onSubmit={submit}
        {...props}
        ref={ref}
      >
        <label className="u-off-screen" htmlFor="search">
          {placeholder || "Search"}
        </label>
        <input
          autoComplete={autocomplete}
          className="p-search-and-filter__input"
          disabled={disabled}
          id="search"
          name="search"
          onChange={(evt) => onChange(evt.target.value)}
          placeholder={placeholder}
          type="search"
          value={value}
        />
        <button
          alt="search"
          className="u-off-screen"
          disabled={disabled}
          type="submit"
        >
          Search
        </button>
      </form>
    );
  }
);

SearchBox.propTypes = {
  autocomplete: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  placeholder: PropTypes.string,
  value: PropTypes.string,
};

SearchBox.displayName = "SearchBox";

export default SearchBox;
