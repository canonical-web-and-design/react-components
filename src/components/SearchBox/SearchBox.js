import classNames from "classnames";
import React from "react";
import PropTypes from "prop-types";

const SearchBox = React.forwardRef(
  (
    {
      autocomplete = "on",
      className,
      disabled,
      externallyControlled,
      onChange,
      onSubmit,
      placeholder = "Search",
      value,
      ...props
    },
    ref
  ) => {
    const input = React.createRef();
    const resetInput = () => {
      onChange && onChange("");
      input.current.value = "";
    };
    const submit = (e) => {
      e.preventDefault();
      onSubmit && onSubmit(e);
    };
    return (
      <form
        className={classNames("p-search-box", className)}
        onSubmit={submit}
        {...props}
        ref={ref}
      >
        <label className="u-off-screen" htmlFor="search">
          {placeholder || "Search"}
        </label>
        <input
          autoComplete={autocomplete}
          className="p-search-box__input"
          disabled={disabled}
          id="search"
          name="search"
          onChange={(evt) => onChange(evt.target.value)}
          placeholder={placeholder}
          ref={input}
          type="search"
          defaultValue={externallyControlled ? undefined : value}
          value={externallyControlled ? value : undefined}
        />
        {value && (
          <button
            alt="reset"
            className="p-search-box__reset"
            disabled={disabled}
            onClick={resetInput}
            type="reset"
          >
            <i className="p-icon--close">Clear search field</i>
          </button>
        )}
        <button
          alt="search"
          className="p-search-box__button"
          disabled={disabled}
          type="submit"
        >
          <i className="p-icon--search">Search</i>
        </button>
      </form>
    );
  }
);

SearchBox.propTypes = {
  autocomplete: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  /**
   * Whether the input value will be controlled via external state.
   */
  externallyControlled: PropTypes.bool,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  placeholder: PropTypes.string,
  value: PropTypes.string,
};

SearchBox.displayName = "SearchBox";

export default SearchBox;
