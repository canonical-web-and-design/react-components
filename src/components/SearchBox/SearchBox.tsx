import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";
import type { HTMLProps, FocusEvent, FormEvent, MouseEvent } from "react";
import { TSFixMe } from "index";

type Props = {
  autocomplete?: "on" | "off";
  className?: string;
  disabled?: boolean;
  externallyControlled?: boolean;
  onClick?: (evt: MouseEvent<HTMLFormElement>) => void;
  onChange?: (evt: TSFixMe) => void;
  onFocus?: (evt: FocusEvent<HTMLFormElement>) => void;
  onSubmit?: (evt: FormEvent) => void;
  placeholder?: string;
  value?: string;
} & HTMLProps<HTMLFormElement>;

const SearchBox = ({
  autocomplete = "on",
  className,
  disabled,
  externallyControlled,
  onChange,
  onSubmit,
  placeholder = "Search",
  value,
  ...formProps
}: Props): JSX.Element => {
  const input = React.createRef<HTMLInputElement>();
  const resetInput = () => {
    onChange && onChange(null);
    input.current.value = "";
  };
  const submit = (evt) => {
    evt.preventDefault();
    onSubmit && onSubmit(evt);
  };
  return (
    <form
      className={classNames("p-search-box", className)}
      onSubmit={submit}
      {...formProps}
    >
      <label className="u-off-screen" htmlFor="search">
        {placeholder || "Search"}
      </label>
      <input
        aria-label="search"
        autoComplete={autocomplete}
        className="p-search-box__input"
        disabled={disabled}
        id="search"
        name="search"
        onChange={(evt) => onChange(evt.target.value)}
        placeholder={placeholder}
        ref={input}
        type="search"
        defaultValue={externallyControlled ? undefined : value}
        value={externallyControlled ? value : undefined}
      />
      {value && (
        <button
          aria-label="reset"
          className="p-search-box__reset"
          disabled={disabled}
          onClick={resetInput}
          type="reset"
        >
          <i className="p-icon--close">Clear search field</i>
        </button>
      )}
      <button
        aria-label="search"
        className="p-search-box__button"
        disabled={disabled}
        type="submit"
      >
        <i className="p-icon--search">Search</i>
      </button>
    </form>
  );
};

SearchBox.propTypes = {
  autocomplete: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  /**
   * Whether the input value will be controlled via external state.
   */
  externallyControlled: PropTypes.bool,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  placeholder: PropTypes.string,
  value: PropTypes.string,
};

export default SearchBox;
