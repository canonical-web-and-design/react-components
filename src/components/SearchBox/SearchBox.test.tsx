import { shallow } from "enzyme";
import React from "react";

import SearchBox from "./SearchBox";

describe("SearchBox ", () => {
  it("renders", () => {
    const wrapper = shallow(<SearchBox />);
    expect(wrapper).toMatchSnapshot();
  });

  it("shows the clear button when there is a value", () => {
    const wrapper = shallow(<SearchBox value="admin" />);
    expect(wrapper.find(".p-search-box__reset").exists()).toBe(true);
  });

  it("can externally control the value", () => {
    const wrapper = shallow(
      <SearchBox externallyControlled onChange={jest.fn()} value="admin" />
    );
    let input = wrapper.find(".p-search-box__input");
    expect(input.prop("defaultValue")).toBe(undefined);
    expect(input.prop("value")).toBe("admin");
    wrapper.setProps({ value: "new-admin" });
    input = wrapper.find(".p-search-box__input");
    expect(input.prop("value")).toBe("new-admin");
  });
});
