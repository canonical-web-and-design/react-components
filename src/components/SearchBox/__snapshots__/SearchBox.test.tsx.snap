// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox  renders 1`] = `
<form
  className="p-search-box"
  onSubmit={[Function]}
>
  <label
    className="u-off-screen"
    htmlFor="search"
  >
    Search
  </label>
  <input
    aria-label="search"
    autoComplete="on"
    className="p-search-box__input"
    id="search"
    name="search"
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <button
    aria-label="search"
    className="p-search-box__button"
    type="submit"
  >
    <i
      className="p-icon--search"
    >
      Search
    </i>
  </button>
</form>
`;
